Optimize glist by avoiding g_list_nth_data

christopherlam opened the pull request:
While exploring alternatives to xaccAccountGetSplitList (gets slow e.g. account has 1000s of splits) I found `g_list_nth_data` being used in a few places which is known to be inefficient.

https://developer.gnome.org/programming-guidelines/stable/glist.html.en

My aim was to traverse an account splitlist by pointer dereferences rather than creating a GList to be converted to SCM list. I am now sure how to do yet. Creating calls to xaccAccountGetFirstSplit, xaccSplitGetNextSplit does not seem currently possible.

But these two optimisations seem very safe and correct.

christopherlam commented on libgnucash/engine/Transaction.c:
```diff
@@ -676,15 +676,23 @@ Transaction *
-    for (i = 0; i < length; ++i)
-	xaccSplitCopyKvp (g_list_nth_data (from->splits, i),
-			    g_list_nth_data (to->splits, i));
+    g_assert (g_list_length (to->splits) == g_list_length (from->splits));
```
Ok repushed. When it passes build I'll merge.

jralls commented:
IIUC what you mean by creating a GList to be converted to a Scheme one, there shouldn't be any need for that: The account's split list is already a GList. Just call `xaccAccountGetSplitList` and traverse it, copying the `Split*`s to your SCM list.

christopherlam commented:
> IIUC what you mean by creating a GList to be converted to a Scheme one, there shouldn't be any need for that: The account's split list is already a GList. Just call `xaccAccountGetSplitList` and traverse it, copying the `Split*`s to your SCM list.

IIUC https://github.com/Gnucash/gnucash/blob/maint/common/base-typemaps.i#L162 already does this for GList -> SCM; creating SCM_list by prepend-and-reverse, and the SCM_list is then processed sequentially afterwards. I was hoping to find a way to convert a GList-type list to a (data . rest) pair struct, with data&rest accessors... but would be very ugly and introduce SCM in Account.cpp.

jralls commented:
`car` and `cdr` just return the list nodes `data` and `next` elements respectively. You could create `gnc_c_list_car` and `gnc_c_list_cdr` that implement that pretty easily and stick them in a convenient `foo.i`.  The only hard part would be keeping track of the type information on whatever pointer `gnc_c_list_car` retrieves.

christopherlam commented:
Here's what I can do so far (currently segfaults) -- it may not be worth the effort though :grin: 

```patch
modified   libgnucash/engine/Account.cpp
@@ -3822,6 +3822,38 @@ xaccAccountGetSplitList (const Account *acc)
     return GET_PRIVATE(acc)->splits;
 }
 
+traversal_s * xaccAccountTraversalSetup (const Account *acc)
+{
+    traversal_s *traversal;
+
+    traversal = g_new0 (traversal_s, 1);
+    traversal->splits = GET_PRIVATE(acc)->splits;
+    return traversal;
+}
+
+Split * xaccAccountTraversalPeek (traversal_s * traversal)
+{
+    SplitList *splits;
+    Split *split;
+
+    g_return_val_if_fail (traversal, NULL);
+
+    splits = traversal->splits;
+    g_return_val_if_fail (splits, NULL);
+
+    split = static_cast<Split*>(splits->data);
+    return split;
+}
+
+Split * xaccAccountTraversalNext (traversal_s * traversal)
+{
+    Split *split;
+
+    split = xaccAccountTraversalPeek (traversal);
+    traversal->splits = traversal->splits->next;
+    return split;
+}
+
 gint64
 xaccAccountCountSplits (const Account *acc, gboolean include_children)
 {
modified   libgnucash/engine/Account.h
@@ -74,6 +74,11 @@ typedef struct
     QofBook *book;
 } GncImportMatchMap;
 
+typedef struct
+{
+    SplitList *splits;
+} traversal_s;
+
 /* --- type macros --- */
 #define GNC_TYPE_ACCOUNT            (gnc_account_get_type ())
 #define GNC_ACCOUNT(o)              \
@@ -1009,6 +1014,9 @@ gboolean xaccAccountIsEquityType(GNCAccountType t);
  */
 SplitList* xaccAccountGetSplitList (const Account *account);
 
+traversal_s * xaccAccountTraversalSetup (const Account *acc);
+Split * xaccAccountTraversalPeek (traversal_s * traversal);
+Split * xaccAccountTraversalNext (traversal_s * traversal);
 
 /** The xaccAccountCountSplits() routine returns the number of all
  *    the splits in the account.
```

jralls commented:
The segfault is probably from running off the end of the list in `xaccAccountTraversalNext`.  You don't show your test code, but remember that you need `g_free(traversal)` somewhere once it runs off the end.

This also doesn't really mirror the lispish `car` and `cdr` because both `xaccAccountTraversalPeek` and `xaccAccountTraversalNext` return the split, which is what `car` does. `traversal_s::splits` is the `cdr`.

Having state on the heap is seriously unFunctional.

The Functional to process a loop is recursion, not iteration:
 ```
(define (for-each bar do-something)
   (if bar
       ((do-something (car bar)
       (for-each (cdr bar) do-something))
```
In C that looks like
```
void
for_each(List* list, CallBack* do_something)
{
    if (list)
    {
        (*do_something) (list->data);
        for_each(list->next, do_something)
     }
}
```
Directly implemented recursion is incredibly inefficient and both lispish languages as well as C compilers from around 1998 (when the first C ISO standard was issued) convert it into a loop of the sort that is any way more natural to C programmers:
```
void
for_each(List* list, Callback* do_something)
{
     for_each(List* node = list; node; node = node->next)
          do_something(node);
}
```
Which is a direct representation of what both lispish and modern procedural languages will generate in machine code for a tail-recursion function like the first two examples.

With that background let's make our C recursive function look a bit more lispish:
```
ListType* item
list_car(List* list)
{
    if (list)
       return list->data;
   return NULL;
}
List*
list_cdr(List* list)
{
   if (list)
       return list->next;
   return NULL;
}

void
for_each(List* list, Callback* do_something)
{
    if (list)
    {
          (*do_something)(list_car(list));
          for_each(list_cdr(list));
    }
}
```

christopherlam commented:
After bashing emacs, and a lot of trial and error, I managed to make it work :roll_eyes:

The struct `splitspair_s` is a lispy `cons` cell, with `car` and `cdr`. The `car` is an obligatory pointer to `Split` or NULL, whereas `cdr` is a `gpointer` which would point at another splitspair_s or NULL. Then I define CAR and CDR accessors. Then I modify `report-utilities.scm` to use the new accessors.

I don't actually know how the cdr accessor manages to find the node->next pointer.

The *main* benefit to this approach is to reduce O(2*N) to O(N) for reports using `gnc:account-accumulate-at-dates`. we are avoiding constantly building lists from GLists every time one is converted from C to SCM, and vice versa. I think this approach copies the account->splits *once* and reuses the same g_list until it is destroyed.

The cost is obviously increased complexity and need for memory management in guile.

Other reports could use this approach too; this includes the price calculators.

I'm not sure that handling of NULL is done 1000% correctly, but all tests are still working well. Ideally the pairs could be generalised to any GList type. I think this approach could be called 'make me a lisp'.

```patch
modified   gnucash/report/report-system/report-utilities.scm
@@ -498,38 +498,49 @@ flawed. see report-utilities.scm. please update reports.")
           (nosplit->elt #f)
           (split->date (compose xaccTransGetDate xaccSplitGetParent))
           (split->elt xaccSplitGetBalance))
-  (let lp ((splits (xaccAccountGetSplitList acc))
+
+  (define original-pair (xaccAccountSplitsPair acc))
+  (let lp ((splits original-pair)
            (dates (sort dates <))
            (result '())
            (last-result nosplit->elt))
     (match dates
 
       ;; end of dates. job done!
-      (() (reverse result))
+      (()
+       (xaccSplitsPairDestroy original-pair)
+       (reverse result))
 
       ((date . rest)
-       (match splits
-
-         ;; end of splits, but still has dates. pad with last-result
-         ;; until end of dates.
-         (() (lp '() rest (cons last-result result) last-result))
-
-         ((head . tail)
-          (let ((next (and (pair? tail) (car tail))))
-            (cond
-
-             ;; the next split is still before date.
-             ((and next (< (split->date next) date))
-              (lp tail dates result (split->elt head)))
-
-             ;; head split after date, accumulate previous result
-             ((< date (split->date head))
-              (lp splits rest (cons last-result result) last-result))
-
-             ;; head split before date, next split after date, or end.
-             (else
-              (let ((head-result (split->elt head)))
-                (lp tail rest (cons head-result result) head-result)))))))))))
+       (cond
+
+        ((null? splits)
+         (lp splits rest (cons last-result result) last-result))
+
+        ;; end of splits, but still has dates. pad with last-result
+        ;; until end of dates.
+        ((null? (xaccSplitsPairCar splits))
+         (lp '() rest (cons last-result result) last-result))
+
+        (else
+         (let* ((head (xaccSplitsPairCar splits))
+                (tail (xaccSplitsPairCdr splits))
+                (next (and (not (null? tail))
+                           (xaccSplitsPairCar tail))))
+           (cond
+
+            ;; the next split is still before date.
+            ((and next (< (split->date next) date))
+             (lp tail dates result (split->elt head)))
+
+            ;; head split after date, accumulate previous result
+            ((< date (split->date head))
+             (lp splits rest (cons last-result result) last-result))
+
+            ;; head split before date, next split after date, or end.
+            (else
+             (let ((head-result (split->elt head)))
+               (lp tail rest (cons head-result result) head-result)))))))))))
 
 ;; This works similar as above but returns a commodity-collector, 
 ;; thus takes care of children accounts with different currencies.
modified   libgnucash/engine/Account.cpp
@@ -3822,6 +3822,68 @@ xaccAccountGetSplitList (const Account *acc)
     return GET_PRIVATE(acc)->splits;
 }
 
+splitspair_s * xaccAccountSplitsPair (const Account *acc)
+{
+    splitspair_s *pair;
+    SplitList *splits;
+
+    pair = g_new0(splitspair_s, 1);
+    splits = xaccAccountGetSplitList (acc);
+    if (splits)
+    {
+        splits = g_list_copy (splits);
+
+        pair->car = static_cast <Split*> (splits->data);
+        pair->cdr = splits->next;
+        if (pair->cdr)
+        {
+            splits->next->prev = NULL;
+            splits->next = NULL;
+            g_list_free (splits);
+        }
+    }
+    else
+    {
+        pair->car = NULL;
+        pair->cdr = NULL;
+    }
+    return pair;
+}
+
+Split * xaccSplitsPairCar (splitspair_s *pair)
+{
+    if (!pair)
+    {
+        PWARN ("argument is not a pair");
+        return NULL;
+    }
+
+    return pair->car;
+}
+
+splitspair_s * xaccSplitsPairCdr (splitspair_s *pair)
+{
+    if (!pair)
+    {
+        PWARN ("argument is not a pair");
+        return NULL;
+    }
+
+    if (pair->car)
+        return (splitspair_s*)pair->cdr;
+
+    return NULL;
+}
+
+void xaccSplitsPairDestroy (splitspair_s *pair)
+{
+    if (pair)
+    {
+        g_free (pair->cdr);
+        g_free (pair);
+    }
+}
+
 gint64
 xaccAccountCountSplits (const Account *acc, gboolean include_children)
 {
modified   libgnucash/engine/Account.h
@@ -74,6 +74,17 @@ typedef struct
     QofBook *book;
 } GncImportMatchMap;
 
+typedef struct
+{
+    SplitList *splits;
+} traversal_s;
+
+typedef struct
+{
+    Split *car;
+    GList *cdr;
+} splitspair_s;
+
 /* --- type macros --- */
 #define GNC_TYPE_ACCOUNT            (gnc_account_get_type ())
 #define GNC_ACCOUNT(o)              \
@@ -1009,6 +1020,10 @@ gboolean xaccAccountIsEquityType(GNCAccountType t);
  */
 SplitList* xaccAccountGetSplitList (const Account *account);
 
+splitspair_s * xaccAccountSplitsPair (const Account *acc);
+Split * xaccSplitsPairCar (splitspair_s *pair);
+splitspair_s * xaccSplitsPairCdr (splitspair_s *pair);
+void xaccSplitsPairDestroy (splitspair_s *pair);
 
 /** The xaccAccountCountSplits() routine returns the number of all
  *    the splits in the account.
```

christopherlam commented:
> reduce ~O(2*N)~ O(3N) [note] to O(N) 

Actually I think it reduces from O(3N) to O(2N), so not really worth the effort here.
Current behaviour, for each account with N splits:
- scan N to build `scm_cons` list in reverse
- scan N list to `scm_reverse`
- use N sequentially in `gnc:account-accumulate-at-dates` (and potentially others)

Lispify the list:
- scan N to build a copy of account->splits with `g_list_copy`
- use N sequentially, with additional overhead of type checking.
